main.cpp

```diff
@@ -71,28 +71,291 @@ write 3 UDTs below that EACH have:
+};
+
+RadioTransmitter::RadioTransmitter() : maxOutputLevel(100.f), minOutputLevel(5.f), antennaImpedance(50.f), microphoneImpedance(4.f), monitorEnabled(true)
+{
+    std::cout << "RadioTransmitter has been constructed." << std::endl;
+}
+
+RadioTransmitter::~RadioTransmitter()
+{
+    std::cout << "RadioTransmitter has been deconstructed." << std::endl;
+}
+
+RadioTransmitter::TransmitterControlInterface::TransmitterControlInterface() : antennaSelection(1), vfoSelection(2), frequency(7.100f), microphoneGain(2.f), mode("AM")
+{
+    std::cout << "TransmitterControlInterface has been constructed." << std::endl;
+}
+
+RadioTransmitter::TransmitterControlInterface::~TransmitterControlInterface()
+{
+    std::cout << "TransmitterControlInterface has been deconstructed." << std::endl;
+}
+
+void RadioTransmitter::enableVoiceMemoryUnit(bool enableVoiceMemoryUnit)
+{
+       std::cout << "Voice Memory Unit Enabled: " << enableVoiceMemoryUnit << std::endl;
+}
+ 
+void RadioTransmitter::enableAntennaTunner(bool enableAntennaTunner)
+{
+       std::cout << "Antenna Tunner Enabled: " << enableAntennaTunner << std::endl;
+}
+
+void RadioTransmitter::enableDataManagementUnit(bool enableDataManagementUnit)
+{
+       std::cout << "Data Management Unit Enabled: " << enableDataManagementUnit << std::endl;
+}
+
+void RadioTransmitter::TransmitterControlInterface::setMode(std::string modeSelection)
+{
+       std::cout << "Tx Mode has been set to: " << modeSelection << std::endl;
+}
+
+void RadioTransmitter::TransmitterControlInterface::txEnabled(bool txEnabled)
+{
+       std::cout << "Tx Enabled: " << txEnabled << std::endl;
+}
+
+void RadioTransmitter::TransmitterControlInterface::enableKeyer(bool keyerEnabled)
+{
+       std::cout << "Keyer Enabled: " << keyerEnabled << std::endl;
+}
+
+
 /*
  copied UDT 3:
  */
+struct PowerSupply
+{
+    PowerSupply();
+    ~PowerSupply();
+
+    float inputPower, outputPower, outputCurrentMax;
+    std::string inputConnectorType, outputConnectorType;
+
+    void enable(bool powerState);
+    void setDisplayMetric(std::string displayMetric);
+    void setInputVoltage(int inputVoltage);
+};
+
+PowerSupply::PowerSupply()
+{
+    std::cout << "PowerSupply has been constructed." << std::endl;
+}
+
+PowerSupply::~PowerSupply()
+{
+    std::cout << "PowerSupply has been deconstructed." << std::endl;
+}
+
+void PowerSupply::enable(bool powerState)
+{
+    std::cout << "PowerSupply Powered On: " << powerState << std::endl;
+}
+
+void PowerSupply::setDisplayMetric(std::string displayMetric)
+{
+    std::cout << "PowerSupply Display Metric: " << displayMetric << std::endl;
+}
+
+void PowerSupply::setInputVoltage(int inputVoltage)
+{
+    std::cout << "PowerSupply Input Voltage: " << inputVoltage << std::endl;
+}
 
 /*
  new UDT 4:
  with 2 member functions
  */
 
+struct Transciever 
+{
+    Transciever();
+    ~Transciever();
+    
+    RadioReceiver rx;
+    RadioTransmitter tx;
+    PowerSupply psu; 
+
+    void powerOn(bool);
```

Comment:
A function with a name like this should not take an argument. 
It is already telling you that it is going to `power on` its component parts.  
If this function was named `setPowerState(PowerState)`, it would make sense to have a function argument:
example: 
```
enum class PowerState : int
{
    Off,  //equivalent to '0'
    On,  //equivalent to '1'
    NO_STATE,
    END_OF_LIST
};

void setPowerState(PowerState newState) { powerState = newState; }

PowerState powerState { PowerState::NO_STATE };
```
- [ ] revise